### PR TITLE
TAS-related: Get client player maxspeed in HL modifications if it's available

### PR DESCRIPTION
**Pull request by SmileyAG**
About factor hack in PARANOIA: https://github.com/FWGS/paranoia_toolkit/blob/b5a073543499f593e8c12bce27840614a5f47d32/code_src/pm_shared/pm_shared.c#L1216-L1217

### CODE REVIEW

File: BunnymodXT/cvars.cpp [hunk 1]
```diff
@@ -11,6 +11,7 @@ namespace CVars
 	CVarWrapper _bxt_bunnysplit_time_update_frequency("_bxt_bunnysplit_time_update_frequency", "41");
 	CVarWrapper _bxt_save_runtime_data_in_demos("_bxt_save_runtime_data_in_demos", "1");
 	CVarWrapper _bxt_tas_script_generation("_bxt_tas_script_generation", "1337");
+	CVarWrapper bxt_tas_clientmaxspeed_in_prediction("bxt_tas_clientmaxspeed_in_prediction", "1");
```
Comment (YaLTeR) [inline BunnymodXT/cvars.cpp:14]: Why is this needed?

File: BunnymodXT/modules/HwDLL.cpp [hunk 2]
```diff
@@ -4733,11 +4746,19 @@ void HwDLL::FindCVarsIfNeeded()
...
 	auto vars = HLStrafe::MovementVars();
+	auto &cl = ClientDLL::GetInstance();
 
 	FindCVarsIfNeeded();
 	vars.Frametime = GetFrameTime();
 	vars.Maxvelocity = CVars::sv_maxvelocity.GetFloat();
-	vars.Maxspeed = CVars::sv_maxspeed.GetFloat();
+
+	if (CVars::bxt_tas_clientmaxspeed_in_prediction.GetBool() && cl.DoesGameDirMatch("paranoia"))
+		vars.Maxspeed = cl.pEngfuncs->GetClientMaxspeed() * 3.2f; // GetMaxSpeed is factor here, 3.2f is approx. multiplier for walkmove
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4756]: Shouldn't it be multiplied by sv_maxspeed?

Comment (SmileyAG) [reply BunnymodXT/modules/HwDLL.cpp:4756]: > Shouldn't it be multiplied by sv_maxspeed?

Got checked rn, and you are right, then formula for getting maxspeed in **PARANOIA** would be like:

`cl.pEngfuncs->GetClientMaxspeed() * CVars::sv_maxspeed.GetFloat() / 100.0f`

File: BunnymodXT/cvars.cpp (hunk 1 above)
Comment (SmileyAG) [reply BunnymodXT/cvars.cpp:14]: Well, with hardcoding clientmaxspeed prediction enabling by default, there is would be no ability to set lower custom maxspeed value in CS games e.g.

Let's say that USP is gives `250` units in CS, but I want to limit it for `200` for some weird reason, then:

- I do `sv_maxspeed 200` and it set lower maxspeed
- But, `GetClientMaxspeed` still goes to output the original value of weapon maxspeed

Comment (YaLTeR) [reply BunnymodXT/cvars.cpp:14]: Then it should take minimum between sv_maxspeed and client maxspeed?

Comment (SmileyAG) [reply BunnymodXT/cvars.cpp:14]: > Then it should take minimum between sv_maxspeed and client maxspeed?

I think u could do more simpler by adding `CVars::sv_maxspeed.GetFloat() > cl.pEngfuncs->GetClientMaxspeed()` condition here: 

https://github.com/YaLTeR/BunnymodXT/blob/6396f62e30088e40ecf06fccb659ea6719223b66/BunnymodXT/modules/HwDLL.cpp#L4757

Comment (YaLTeR) [reply BunnymodXT/cvars.cpp]: You should double check what actually happens in the movement code for different client maxspeed vs. sv_maxspeed

File: BunnymodXT/modules/HwDLL.cpp [hunk 3]
```diff
@@ -4733,11 +4745,19 @@ void HwDLL::FindCVarsIfNeeded()
...
 {
 	auto vars = HLStrafe::MovementVars();
+	auto &cl = ClientDLL::GetInstance();
 
 	FindCVarsIfNeeded();
 	vars.Frametime = GetFrameTime();
 	vars.Maxvelocity = CVars::sv_maxvelocity.GetFloat();
-	vars.Maxspeed = CVars::sv_maxspeed.GetFloat();
+
+	if (cl.DoesGameDirMatch("paranoia"))
```
Comment (YaLTeR) [inline BunnymodXT/modules/HwDLL.cpp:4754]: Can we cache this somehow? Like with a `static bool is_paranoia`. Because it's a string comparison every frame, which is expensive.

**Review by YaLTeR — COMMENTED** [review]
This should be good to merge otherwise right?

**Comment by YaLTeR** [comment]
Thanks
